File: pippip-native-ios/Support/ContactManager.m
//
//  ContactManager.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 12/12/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "ContactManager.h"
#import "pippip_native_ios-Swift.h"
#import "ApplicationSingleton.h"
#import "LoggingErrorDelegate.h"
#import "RESTSession.h"
#import "EnclaveRequest.h"
#import "EnclaveResponse.h"
#import "ContactDatabase.h"
#import "Notifications.h"

typedef enum REQUEST { MATCH_NICKNAME, ADD_FRIEND, DELETE_FRIEND, REQUEST_CONTACT,
                        ACKNOWLEDGE_REQUEST, GET_REQUESTS, DELETE_CONTACT, UPDATE_POLICY,
                        SYNC_CONTACTS, UPDATE_NICKNAME, UPDATE_PENDING_CONTACTS, NONE
} ContactRequest;

@interface ContactManager ()
{
    ContactRequest contactRequest;
    NSMutableArray<Contact*> *contactList;
    NSMutableDictionary<NSString*, Contact*> *contactMap;
    NSString *requestedNickname;
    NSString *acknowledgedNickname;
}

@property (weak, nonatomic) RESTSession *session;
@property (weak, nonatomic) ContactDatabase *contactDatabase;

@end

@implementation ContactManager

@synthesize errorDelegate;
@synthesize postPacket;

- (instancetype)init {
    self = [super init];
    
    _session = nil;
    _contactDatabase = [ApplicationSingleton instance].contactDatabase;
    contactMap = [NSMutableDictionary dictionary];
    contactRequest = NONE;
    errorDelegate = [[NotificationErrorDelegate alloc] initWithTitle:@"Contact Error"];

    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(contactsUpdated:)
                                               name:CONTACTS_UPDATED object:nil];
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(newSession:)
                                               name:NEW_SESSION object:nil];
    return self;
    
}

- (void)dealloc {

    [[NSNotificationCenter defaultCenter] removeObserver:self name:CONTACTS_UPDATED object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:NEW_SESSION object:nil];

}

- (void)acknowledgeRequest:(NSString *)response withId:(NSString *)publicId withNickname:(NSString*) nickname {

    acknowledgedNickname = nickname;
    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"AcknowledgeRequest";
    request[@"id"] = publicId;
    request[@"response"] = response;
    [self sendRequest:request with:ACKNOWLEDGE_REQUEST];
    
}

- (BOOL)addFriend:(NSString *)publicId {

    NSInteger found = [[ApplicationSingleton instance].config whitelistIndexOf:publicId];
    if (found == NSNotFound) {
        NSMutableDictionary *request = [NSMutableDictionary dictionary];
        request[@"method"] = @"UpdateWhitelist";
        request[@"action"] = @"add";
        request[@"id"] = publicId;
        [self sendRequest:request with:ADD_FRIEND];
        return YES;
    }
    else {
        return NO;
    }

}

- (void)contactsUpdated:(NSNotification*)notification {

    contactList = [[_contactDatabase getContactList] mutableCopy];
    [self mapContacts];

}

- (NSArray*)decodeKeys:(NSArray*)keyStrings {
    
    NSMutableArray *keys = [NSMutableArray array];
    for (NSString *key in keyStrings) {
        [keys addObject:[[NSData alloc] initWithBase64EncodedString:key options:0]];
    }
    return keys;
    
}

- (void)deleteContact:(NSString *)publicId {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"DeleteContact";
    request[@"publicId"] = publicId;
    [self sendRequest:request with:DELETE_CONTACT];
    
}

- (void)deleteFriend:(NSString *)publicId {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"UpdateWhitelist";
    request[@"action"] = @"delete";
    request[@"id"] = publicId;
    [self sendRequest:request with:DELETE_FRIEND];
    
}

- (void)dispatchResponse:(NSDictionary*)response {

    switch (contactRequest) {
        case ADD_FRIEND:
            NSLog(@"Add friend result - %@", response[@"result"]);
            [AsyncNotifier notifyWithName:FRIEND_ADDED object:nil userInfo:response];
            break;
        case DELETE_FRIEND:
            NSLog(@"Delete friend result - %@", response[@"result"]);
            [AsyncNotifier notifyWithName:FRIEND_DELETED object:nil userInfo:response];
            break;
        case DELETE_CONTACT:
        {
            [[ApplicationSingleton instance].conversationCache deleteAllMessages:response[@"publicId"]];
            NSArray *toDelete = [NSArray arrayWithObjects:response[@"publicId"], nil];
            [_contactDatabase deleteContacts:toDelete];
            contactList = [[_contactDatabase getContactList] mutableCopy];
            [AsyncNotifier notifyWithName:CONTACT_DELETED object:nil userInfo:response];
            [AsyncNotifier notifyWithName:MESSAGES_UPDATED object:nil userInfo:nil];
        }
            break;
        case MATCH_NICKNAME:
            [AsyncNotifier notifyWithName:NICKNAME_MATCHED object:nil userInfo:response];
            break;
        case REQUEST_CONTACT:
        {
            NSMutableDictionary *convert = [NSMutableDictionary dictionary];
            convert[@"publicId"] = response[@"requestedContactId"];
            convert[@"status"] = response[@"result"];
            if (requestedNickname != nil) {
                convert[@"nickname"] = requestedNickname;
            }
            Contact *contact = [[Contact alloc] init:convert];
            [_contactDatabase addContact:contact];  // Sends contacts updated notification
            [AsyncNotifier notifyWithName:CONTACT_REQUESTED object:contact userInfo:nil];
        }
            break;
        case UPDATE_NICKNAME:
            [AsyncNotifier notifyWithName:NICKNAME_UPDATED object:nil userInfo:response];
            break;
        case UPDATE_POLICY:
            [AsyncNotifier notifyWithName:POLICY_UPDATED object:nil userInfo:response];
            break;
        case GET_REQUESTS:
        {
            NSArray *requests = response[@"requests"];
            if (requests.count > 0) {
                [AsyncNotifier notifyWithName:REQUESTS_UPDATED object:requests userInfo:nil];
            }
        }
            break;
        case SYNC_CONTACTS:
        {
            NSInteger added = [response[@"added"] integerValue];
            NSInteger deleted = [response[@"deleted"] integerValue];
            NSInteger updated = [response[@"updated"] integerValue];
            NSLog(@"Contacts synchronized, %ld added, %ld updated, %ld deleted", added, updated, deleted);
            [AsyncNotifier notifyWithName:CONTACTS_SYNCRHONIZED object:nil userInfo:response];
        }
            break;
        case ACKNOWLEDGE_REQUEST:
        {
            NSDictionary *contact = response[@"acknowledged"];
            Contact *translated = [self translateContact:contact];
            if (translated != nil && [translated.status isEqualToString:@"accepted"]) {
                [_contactDatabase addContact:translated];
                [self loadContactList];
            }
//            NSArray *pending = response[@"pending"];
            NSArray *pending = [NSArray array];
            [AsyncNotifier notifyWithName:REQUEST_ACKNOWLEDGED object:pending userInfo:nil];
        }
            break;
        case UPDATE_PENDING_CONTACTS:
        {
            NSArray *contacts = response[@"contacts"];
            if (contacts != nil && contacts.count > 0)
            [self updateContacts:contacts];
            [AsyncNotifier notifyWithName:CONTACTS_UPDATED object:nil userInfo:nil];
        }
            break;
        case NONE:
            break;
    }

}

- (Contact*)getContact:(NSString *)publicId {

    return contactMap[publicId];

}

- (Contact*)getContactById:(NSInteger)contactId {

    [self loadContactList];
    for (Contact *contact in contactList) {
        if (contact.contactId == contactId) {
            return contact;
        }
    }
    return nil;

}

- (NSArray*)getContactList {

    [self loadContactList];
    return contactList;

}

- (NSString*)getContactPublicId:(NSString*)nickname {

    for (Contact *contact in contactList) {
        if ([contact.nickname isEqualToString:nickname]) {
            return contact.publicId;
        }
    }
    return nil;

}

- (NSArray<Contact*>*)getContacts:(NSString *)status {
    
    NSMutableArray *filtered = [NSMutableArray array];
    for (Contact *contact in contactList) {
        NSString *currentStatus = contact.status;
        if ([status isEqualToString:currentStatus]) {
            [filtered addObject:contact];
        }
    }
    return filtered;
    
}

- (void)getRequests {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"GetPendingRequests";
    [self sendRequest:request with:GET_REQUESTS];
    
}

- (void)loadContactList {
    
    if (contactList == nil) {
        contactList = [[_contactDatabase getContactList] mutableCopy];
        [self mapContacts];
    }
    
}

- (void)mapContacts {

    [contactMap removeAllObjects];
    for (Contact *contact in contactList) {
        contactMap[contact.publicId] = contact;
    }

}

- (void)matchNickname:(NSString *)nickname withPublicId:(NSString *)publicId{

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"MatchNickname";
    if (nickname != nil) {
        request[@"nickname"] = nickname;
    }
    if (publicId != nil) {
        request[@"publicId"] = publicId;
    }
    [self sendRequest:request with:MATCH_NICKNAME];
    
}

- (void)newSession:(NSNotification*)notification {

    contactList = [[_contactDatabase getContactList] mutableCopy];
    [self mapContacts];

}

- (void)postComplete:(NSDictionary*)response {

    if (response != nil) {
        EnclaveResponse *enclaveResponse = [[EnclaveResponse alloc] init];
        if ([enclaveResponse processResponse:response errorDelegate:errorDelegate]) {
            NSDictionary *contactResponse = [enclaveResponse getResponse];
            [self dispatchResponse:contactResponse];
        }
    }
    contactRequest = NONE;

}

- (void)requestContact:(NSString*)publicId withNickname:(NSString *)nickname {

    requestedNickname = nickname;
    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"RequestContact";
    request[@"id"] = publicId;
    [self sendRequest:request with:REQUEST_CONTACT];
    
}

/*
 * fragment should be upper case
 */
- (NSArray<Contact*>*)searchContacts:(NSString *)fragment {

    NSMutableArray *results = [NSMutableArray array];
    for (Contact *contact in contactList) {
        NSString *nickname = nil;
        if (contact.nickname != nil) {
            nickname = [contact.nickname uppercaseString];
        }
        NSString *publicId = [contact.publicId uppercaseString];
        if ([publicId containsString:fragment]) {
            [results addObject:contact];
        }
        else if (nickname != nil) {
            if ([nickname containsString:fragment]) {
                [results addObject:contact];
            }
        }
    }
    
    return results;

}

- (void)sendRequest:(NSDictionary*)request with:(ContactRequest)type {

    if (contactRequest == NONE) {
        contactRequest = type;
        if (_session == nil) {
            _session = [ApplicationSingleton instance].restSession;
        }
        
        EnclaveRequest *enclaveRequest = [[EnclaveRequest alloc] init];
        [enclaveRequest setRequest:request];
        
        postPacket = enclaveRequest;
        [_session queuePost:self];
    }
    else {
        NSLog(@"Contact manager request out of order");
    }
    
}

- (void)sessionComplete:(NSDictionary*)response {
    // Nothing to do here. Session is already established.
}

- (void)setContactPolicy:(NSString*)policy {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"SetContactPolicy";
    request[@"policy"] = policy;
    [self sendRequest:request with:UPDATE_POLICY];

}

- (void)syncContacts {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    NSArray *contactList = [_contactDatabase getContactList];
    NSMutableArray *syncList = [NSMutableArray array];
    for (Contact *contact in contactList) {
        Contact *sync = [[Contact alloc] init];
        sync.publicId = contact.publicId;
        sync.status = contact.status;
        sync.currentSequence = contact.currentSequence;
        sync.currentIndex = contact.currentIndex;
        sync.timestamp = contact.timestamp;
        [syncList addObject:sync];
    }
    request[@"method"] = @"SyncContacts";
    request[@"contacts"] = syncList;
    [self sendRequest:request with:SYNC_CONTACTS];

}

- (Contact*)translateContact:(NSDictionary*)serverContact {

    NSString *publicId = serverContact[@"publicId"];
    NSString *status = serverContact[@"status"];
    NSString *authData = serverContact[@"authData"];
    NSString *nonce = serverContact[@"nonce"];
    NSArray *keyStrings = serverContact[@"messageKeys"];
    if (publicId == nil || status == nil || authData == nil || keyStrings == nil || nonce == nil) {
        NSLog(@"Invalid server response in acknowledged contact");
        return nil;
    }
    else if ([status isEqualToString:@"accepted"]) {
        Contact *entity = [[Contact alloc] init];
        entity.publicId = publicId;
        if (acknowledgedNickname != nil) {
            entity.nickname = acknowledgedNickname;
        }
        entity.currentIndex = 0;
        entity.currentSequence = 0;
        entity.timestamp = [serverContact[@"timestamp"] integerValue];
        entity.status = status;
        NSData *adBytes = [[NSData alloc] initWithBase64EncodedString:authData options:0];
        NSData *nonceBytes = [[NSData alloc] initWithBase64EncodedString:nonce options:0];
        if (adBytes != nil && nonceBytes != nil) {
            entity.authData = adBytes;
            entity.nonce = nonceBytes;
            NSArray *keys = [self decodeKeys:keyStrings];
            if (keys != nil) {
                entity.messageKeys = keys;
                return entity;
            }
        }
        else {
            NSLog(@"Encoding error in translate contact");
        }
    }
    return nil;

}

- (void)updateContacts:(NSArray<NSDictionary*>*)serverContacts {
    
    NSMutableArray *updates = [NSMutableArray array];
    for (NSDictionary *serverContact in serverContacts) {
        NSString *publicId = serverContact[@"publicId"];
        Contact *update = contactMap[publicId];
        if (update == nil) {
            // Something really wrong here
            NSLog(@"Process contact, contact %@ does not exist", publicId);
        }
        else {
            NSString *status = serverContact[@"status"];
            update.status = status;
            update.timestamp = [serverContact[@"timestamp"] integerValue];
            if ([status isEqualToString:@"accepted"]) {
                update.currentSequence = 0;
                update.currentIndex = 0;
                update.authData = [[NSData alloc] initWithBase64EncodedString:serverContact[@"authData"] options:0];
                update.nonce = [[NSData alloc] initWithBase64EncodedString:serverContact[@"nonce"] options:0];
                NSArray *messageKeys = serverContact[@"messageKeys"];
                NSMutableArray *keys = [NSMutableArray array];
                for (NSString *keyString in messageKeys) {
                    NSData *key = [[NSData alloc] initWithBase64EncodedString:keyString options:0];
                    [keys addObject:key];
                }
                update.messageKeys = keys;
            }
            [updates addObject:update];
        }
    }
    [_contactDatabase updateContacts:updates];

}

- (void)updateNickname:(NSString *)nickname withOldNickname:(NSString *)oldNickname {

    NSMutableDictionary *request = [NSMutableDictionary dictionary];
    request[@"method"] = @"SetNickname";
    if (nickname != nil) {
        request[@"newNickname"] = nickname;
    }
    else {
        request[@"newNickname"] = @"";
    }
    if (oldNickname != nil) {
        request[@"oldNickname"] = oldNickname;
    }
    else {
        request[@"oldNickname"] = @"";
    }
    [self sendRequest:request with:UPDATE_NICKNAME];
    
}

- (NSUInteger)updatePendingContacts {

    NSMutableArray *pending = [NSMutableArray array];
    NSArray *filtered = [self getContacts:@"pending"];
    for (Contact *entity in filtered) {
        [pending addObject:entity.publicId];
    }
    if (pending.count > 0) {
        NSLog(@"%@", @"Updating pending contacts");
        NSMutableDictionary *request = [NSMutableDictionary dictionary];
        request[@"method"] = @"UpdatePendingContacts";
        request[@"pending"] = pending;
        [self sendRequest:request with:UPDATE_PENDING_CONTACTS];
    }
    return pending.count;

}

@end

File: pippip-native-ios/Support/ContactManager.swift
//
//  ContactManager.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 4/10/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import RKDropdownAlert
import ChameleonFramework

struct ContactRequest: Hashable {

    var nickname: String?
    var publicId: String
    var hashValue: Int {
        return publicId.hashValue
    }

    init(publicId: String, nickname: String?) {

        self.publicId = publicId
        self.nickname = nickname

    }

    init?(serverRequest: [AnyHashable: Any]) {

        guard let puid = serverRequest["publicId"] as? String else { return nil }
        publicId = puid
        nickname = serverRequest["nickname"] as? String

    }

    static func ==(lhs: ContactRequest, rhs: ContactRequest) -> Bool {
        return lhs.publicId == rhs.publicId
    }

}

class ContactManager: NSObject {

    static private var contactList = [Contact]()
    static private var contactMap = [String: Contact]()
    static private var contactIdMap = [Int: Contact]()
    static private var requestSet = Set<ContactRequest>()
    static private var initialized = false

    var contactDatabase: ContactDatabase
    var config = Configurator()
    var pendingRequests: Set<ContactRequest> {
        return ContactManager.requestSet
    }
 
    override init() {

        contactDatabase = ContactDatabase()

        super.init()

        let sessionState = SessionState()
        if (sessionState.authenticated && !ContactManager.initialized) {
            ContactManager.initialized = true
            ContactManager.contactList = contactDatabase.getContactList()
            mapContacts()
        }
        
    }

    func acknowledgeRequest(contactRequest: ContactRequest, response: String) {

        var request = [AnyHashable: Any]()
        request["method"] = "AcknowledgeRequest"
        request["id"] = contactRequest.publicId
        request["response"] = response
        let ackTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            if let serverContact = response["acknowledged"] as? [AnyHashable: Any],
                let acknowledged = ContactRequest(serverRequest: serverContact),
                acknowledged == contactRequest,
                let contact = Contact(serverContact: serverContact) {
                contact.nickname = contactRequest.nickname
                self.addContact(contact)
                ContactManager.requestSet.remove(contactRequest)
                DispatchQueue.global().async {
                    NotificationCenter.default.post(name: Notifications.RequestAcknowledged, object: nil)
                }
            }
            else {
                var info = [AnyHashable: Any]()
                info["title"] = "Contact Error"
                info["message"] = "Invalid response to acknowledgement"
                NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            }
        })
        ackTask.errorTitle = "Contact Error"
        ackTask.sendRequest(request)

    }

    func addContact(_ contact: Contact) {

        if ContactManager.contactMap[contact.publicId] != nil {
            var info = [AnyHashable:Any]()
            info["title"] = "Contact Error"
            info["message"] = "Attempted to add duplicate contact"
            NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
        }
        else {
            contact.contactId = config.newContactId()
            contactDatabase.add(contact)
            ContactManager.contactMap[contact.publicId] = contact
            ContactManager.contactIdMap[contact.contactId] = contact
            ContactManager.contactList.append(contact)
        }
        
    }
    
    func addFriend(_ publicId:String) -> Bool {

        let found = config.whitelistIndex(of: publicId)
        if found == NSNotFound {
            var request = [AnyHashable: Any]()
            request["method"] = "UpdateWhitelist"
            request["id"] = publicId
            request["action"] = "add"
            let addTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
                NotificationCenter.default.post(name: Notifications.FriendAdded, object: nil, userInfo: response)
            })
            addTask.errorTitle = "Friends List Error"
            addTask.sendRequest(request)
            return true
        }
        else {
            return false
        }

    }

    // Returns true if new requests were added
    func addRequests(_ requests: [[AnyHashable: Any]]) {

        for request in requests {
            if let contactRequest = ContactRequest(serverRequest: request) {
                ContactManager.requestSet.insert(contactRequest)
            }
        }

    }

    func allContactIds() -> [Int] {

        var allIds = [Int]()
        for contact in ContactManager.contactList {
            allIds.append(contact.contactId)
        }
        return allIds

    }

    @objc func clearContacts() {

        ContactManager.contactList.removeAll()
        ContactManager.contactMap.removeAll()
        ContactManager.contactIdMap.removeAll()
        ContactManager.initialized = false

    }

    func deleteContact(_ publicId: String) {
        
        var request = [AnyHashable: Any]()
        request["method"] = "DeleteContact"
        request["publicId"] = publicId
        let deleteTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            if let publicId = response["publicId"] as? String {
                // If there are duplicates in the database, this will prevent crashes
                // However, duplicates will require two separate deletes
                if let contactId = ContactManager.contactMap[publicId]?.contactId {
                    self.contactDatabase.deleteContact(contactId)
                    ContactManager.contactMap.removeValue(forKey: publicId)
                    let messageDatabase = MessagesDatabase()
                    messageDatabase.clearMessages(contactId)
                    var newList = [Contact]()
                    for contact in ContactManager.contactList {
                        if contact.publicId != publicId {
                            newList.append(contact)
                        }
                    }
                    ContactManager.contactList = newList
                    NotificationCenter.default.post(name: Notifications.ContactDeleted, object: publicId, userInfo: nil)
                    NotificationCenter.default.post(name: Notifications.MessagesUpdated, object: nil)
                }
            }
        })
        deleteTask.errorTitle = "Contact Error"
        deleteTask.sendRequest(request)
        
    }

    func deleteFriend(_ publicId: String) {

        var request = [AnyHashable: Any]()
        request["method"] = "UpdateWhitelist"
        request["id"] = publicId
        request["action"] = "delete"
        let deleteTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
                NotificationCenter.default.post(name: Notifications.FriendDeleted, object: nil, userInfo: response)
        })
        deleteTask.errorTitle = "Friends List Error"
        deleteTask.sendRequest(request)

    }
    
    @objc func getContact(_ publicId: String) -> Contact? {

        return ContactManager.contactMap[publicId]

    }

    @objc func getContactById(_ contactId: Int) -> Contact? {

        return ContactManager.contactIdMap[contactId]

    }

    func getContactId(_ publicId: String) -> Int {

        if let contact = ContactManager.contactMap[publicId] {
            return contact.contactId
        }
        else {
            return NSNotFound
        }

    }

    @objc func getContactList() -> [Contact] {
        return ContactManager.contactList
    }

    @objc func getPendingRequests() {

        var request = [AnyHashable: Any]()
        request["method"] = "GetPendingRequests"
        let getTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            if let requests = response["requests"] as? [[AnyHashable: Any]] {
                print("\(requests.count) pending requests returned")
                if requests.count > 0 {
                    self.addRequests(requests)
                }
                else {
                    ContactManager.requestSet.removeAll()
                }
                DispatchQueue.global().async {
                    NotificationCenter.default.post(name: Notifications.RequestsUpdated,
                                                    object: ContactManager.requestSet.count)
                }
            }
            else {
                var info = [AnyHashable: Any]()
                info["title"] = "Contact Error"
                info["message"] = "Invalid response to get requests"
                NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            }
        })
        getTask.errorTitle = "Contact Error"
        getTask.sendRequest(request)
        
    }

    func mapContacts() {
        
        ContactManager.contactMap.removeAll()
        for contact in ContactManager.contactList {
            ContactManager.contactMap[contact.publicId] = contact
            ContactManager.contactIdMap[contact.contactId] = contact
        }
        
    }
    
    @objc func matchNickname(nickname: String?, publicId: String?) {

        var request = [AnyHashable: Any]()
        request["method"] = "MatchNickname"
        if (nickname != nil) {
            request["nickname"] = nickname
        }
        if (publicId != nil) {
            request["publicId"] = publicId
        }
        let matchTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            NotificationCenter.default.post(name: Notifications.NicknameMatched, object: nil, userInfo: response)
        })
        matchTask.errorTitle = "Nickname Error"
        matchTask.sendRequest(request)
        
    }

    func requestContact(publicId: String, nickname: String?) {

        var request = [AnyHashable: Any]()
        request["method"] = "RequestContact"
        request["id"] = publicId
        if let _ = ContactManager.contactMap[publicId] {
            var info = [AnyHashable: Any]()
            info["title"] = "Contact Error"
            info["message"] = "This contact already exists in your contact list"
            NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
        }
        else {
            let reqTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
                let contact = Contact()
                contact.publicId = response["requestedContactId"] as! String
                contact.nickname = nickname
                contact.status = response["result"] as! String
                self.addContact(contact)
                NotificationCenter.default.post(name: Notifications.ContactRequested, object: contact)
            })
            reqTask.errorTitle = "Contact Error"
            reqTask.sendRequest(request)
        }
        
    }

    func searchContacts(_ fragment:String) -> [Contact] {

        var result = [Contact]()
        for contact in ContactManager.contactList {
            let pCompare = contact.publicId.uppercased()
            if (pCompare.contains(fragment)) {
                result.append(contact)
            }
            else if let nickname = contact.nickname {
                let nCompare = nickname.uppercased()
                if (nCompare.contains(fragment)) {
                    result.append(contact)
                }
            }
        }
        return result

    }

    @objc func setContactPolicy(_ policy: String) {
        
        var request = [AnyHashable: Any]()
        request["method"] = "SetContactPolicy"
        request["policy"] = policy
        let setTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            NotificationCenter.default.post(name: Notifications.PolicyUpdated, object: nil, userInfo: response)
        })
        setTask.errorTitle = "Policy Error"
        setTask.sendRequest(request)
        
    }

    @objc func updateNickname(newNickname: String?, oldNickname: String?) {

        var request = [AnyHashable: Any]()
        request["method"] = "SetNickname"
        if (oldNickname != nil) {
            request["oldNickname"] = oldNickname
        }
        else {
            request["oldNickname"] = ""
        }
        if (newNickname != nil) {
            request["newNickname"] = newNickname
        }
        else {
            request["newNickname"] = ""
        }
        let setTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            NotificationCenter.default.post(name: Notifications.NicknameUpdated, object: nil, userInfo: response)
        })
        setTask.errorTitle = "Nickname Error"
        setTask.sendRequest(request)
        
    }
    
    func updateContact(_ contact: Contact) {

        contactDatabase.update([contact])
        ContactManager.contactMap[contact.publicId] = contact
        for index in 0..<ContactManager.contactList.count {
            if ContactManager.contactList[index].publicId == contact.publicId {
                ContactManager.contactList[index] = contact
            }
        }

    }

    func updateContacts(_ serverContacts: [[AnyHashable: Any]]) -> [Contact] {

        var updates = [Contact]()

        for serverContact in serverContacts {
            let publicId = serverContact["publicId"] as? String ?? ""
            if let contact = ContactManager.contactMap[publicId] {
                contact.status = serverContact["status"] as! String
                contact.timestamp = (serverContact["timestamp"] as! NSNumber).int64Value
                let authData = serverContact["authData"] as! String
                contact.authData = Data(base64Encoded: authData)
                let nonce = serverContact["nonce"] as! String
                contact.nonce = Data(base64Encoded: nonce)
                let keys = serverContact["messageKeys"] as! [String]
                var messageKeys = [Data]()
                for key in keys {
                    messageKeys.append(Data(base64Encoded: key)!)
                }
                contact.messageKeys = messageKeys
                updates.append(contact)
            }
            else {
                var info = [AnyHashable: Any]()
                info["title"] = "Contact Error"
                info["message"] = "Updated contact not found"
                NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            }
        }
        contactDatabase.update(updates)
        return updates

    }

    @objc func updatePendingContacts() {

        var pending = [String]()
        for contact in ContactManager.contactList {
            if contact.status == "pending" {
                pending.append(contact.publicId)
            }
        }

        var request = [AnyHashable: Any]()
        request["method"] = "UpdatePendingContacts"
        request["pending"] = pending
        let updateTask = EnclaveTask({ (response: [AnyHashable: Any]) -> Void in
            if let serverContacts = response["contacts"] as? [[AnyHashable: Any]] {
                if serverContacts.count > 0 {
                    let updated = self.updateContacts(serverContacts)
                    NotificationCenter.default.post(name: Notifications.PendingContactsUpdated, object: updated)
                    print("\(updated.count) contacts updated")
                }
            }
        })
        updateTask.errorTitle = "Contact Error"
        updateTask.sendRequest(request)
        
    }

}

// Debug only
extension ContactManager {

    func deleteServerContact(nickname: String) {

        NotificationCenter.default.addObserver(self, selector: #selector(nicknameMatched(_:)),
                                               name: Notifications.NicknameMatched, object: nil)
        
        matchNickname(nickname: nickname, publicId: nil)

    }

    @objc func nicknameMatched(_ notification: Notification) {

        NotificationCenter.default.removeObserver(self, name: Notifications.NicknameMatched, object: nil)
        if let info = notification.userInfo,
            let publicId = info["publicId"] as? String {
            deleteContact(publicId)
        }
        else {
            let alertColor = UIColor.flatSand
            RKDropdownAlert.title("Delete Contact Error", message: "That nickname doesn't exist",
                                  backgroundColor: alertColor,
                                  textColor: ContrastColorOf(alertColor, returnFlat: true),
                                  time: 2, delegate: nil)
        }

    }

}
